### PR TITLE
FRAME: utility dispatch_as_account call

### PR DESCRIPTION
**Pull request by muharem**
Introduce `dispatch_as_account` call to utility pallet.

The call useful when calls from the batch needs to be commanded by some pallet's origin (e.g. permissioned schedule call), and the inner call (e.g. transfer) should be dispatched from a sovereign account (as signed).
Real use case: https://github.com/paritytech/polkadot-sdk/pull/2939;

Example:
In the example scheduler schedule operation is permissioned 
``` 
xcm::Transact {
  origin_kind: Xcm / SovereignAccount 
  call: batch_all(
    pallet_scheduler(
      schedule(
        ...
        call: transfer(from: Treasurer, to: Alice, ...)
      )
    ),
  transfer(from: Treasurer, to: Alice, ... )
  )
}
```
Possible Solutions:
1. origin_kind = SovereignAccount && white list `Treasurer` sovereign account for schedule.
2. origin_kind = Xcm && use something like `dispatch_as_account`

I am in favor of (1) solution. But I am keeping this PR for discussions for now.

### CODE REVIEW

File: substrate/frame/utility/src/lib.rs [hunk 1]
```diff
@@ -93,18 +97,23 @@ pub mod pallet {
...
 
-		/// The caller origin, overarching type of all pallets origins.
-		type PalletsOrigin: Parameter +
-			Into<<Self as frame_system::Config>::RuntimeOrigin> +
-			IsType<<<Self as frame_system::Config>::RuntimeOrigin as frame_support::traits::OriginTrait>::PalletsOrigin>;
+		/// The type to convert a pallet's origin to it's sovereign account ID used to dispatch a
+		/// target call with `dispatch_as_account`.
+		///
+		/// This should return an account ID for `Root` origin in benchmarks environment in order to
+		/// calculated the weight of `dispatch_as_account`. If the `Root` is not convertible the
```
Comment (georgepisaltu) [inline substrate/frame/utility/src/lib.rs:104]:
```suggestion
		/// calculate the weight of `dispatch_as_account`. If the `Root` is not convertible, the
```

File: substrate/frame/utility/src/lib.rs [hunk 2]
```diff
@@ -93,18 +97,23 @@ pub mod pallet {
 			+ IsSubType<Call<Self>>
 			+ IsType<<Self as frame_system::Config>::RuntimeCall>;
 
-		/// The caller origin, overarching type of all pallets origins.
-		type PalletsOrigin: Parameter +
-			Into<<Self as frame_system::Config>::RuntimeOrigin> +
-			IsType<<<Self as frame_system::Config>::RuntimeOrigin as frame_support::traits::OriginTrait>::PalletsOrigin>;
+		/// The type to convert a pallet's origin to it's sovereign account ID used to dispatch a
```
Comment (georgepisaltu) [inline substrate/frame/utility/src/lib.rs:100]: Sovereign account associated with a pallet origin feels like wrong phrasing, but I'm not sure.

```suggestion
		/// The type to convert a pallet's origin to an account ID used to dispatch a
```

File: substrate/frame/utility/src/lib.rs [hunk 3]
```diff
@@ -93,18 +97,23 @@ pub mod pallet {
...
+		/// target call with `dispatch_as_account`.
+		///
+		/// This should return an account ID for `Root` origin in benchmarks environment in order to
+		/// calculated the weight of `dispatch_as_account`. If the `Root` is not convertible the
+		/// call is weightless.
+		/// If `Root` is not supposed to have an account for you runtime, use a separate type under
```
Comment (georgepisaltu) [inline substrate/frame/utility/src/lib.rs:106]:
```suggestion
		/// If `Root` shouldn't have an associated account ID, use a separate type under
```

File: substrate/frame/utility/src/lib.rs [hunk 4]
```diff
@@ -491,6 +502,41 @@ pub mod pallet {
 			let res = call.dispatch_bypass_filter(frame_system::RawOrigin::Root.into());
 			res.map(|_| ()).map_err(|e| e.error)
 		}
+
+		/// Dispatches a function call with a sovereign account of the origin.
```
Comment (georgepisaltu) [inline substrate/frame/utility/src/lib.rs:506]: Again "sovereign" account feels off.

File: substrate/frame/utility/src/benchmarking.rs [hunk 5]
```diff
@@ -70,7 +73,7 @@ benchmarks! {
 		let call = Box::new(frame_system::Call::remark { remark: vec![] }.into());
 		let origin: T::RuntimeOrigin = RawOrigin::Signed(caller).into();
 		let pallets_origin: <T::RuntimeOrigin as frame_support::traits::OriginTrait>::PalletsOrigin = origin.caller().clone();
-		let pallets_origin = Into::<T::PalletsOrigin>::into(pallets_origin);
+		// let pallets_origin = Into::<T::PalletsOrigin>::into(pallets_origin);
```
Comment (georgepisaltu) [inline substrate/frame/utility/src/benchmarking.rs:76]: Will remove?

**Review by georgepisaltu — APPROVED** [review]
LGTM

**Comment by xlc** [comment]
Just dispatch using signed origin? This requires all the chains to add this new config event if we don’t use it at all.

**Comment by muharem** [comment]
> Just dispatch using signed origin? This requires all the chains to add this new config event if we don’t use it at all.

I am not happy with current solution as well. I probably close this PR and go for the (1) solution described below.

Example:
In the example scheduler schedule operation is permissioned 
``` 
xcm::Transact {
  origin_kind: Xcm / SovereignAccount 
  call: batch_all(
    pallet_scheduler(
      schedule(
        ...
        call: transfer(from: Treasurer, to: Alice, ...)
      )
    ),
  transfer(from: Treasurer, to: Alice, ... )
  )
}
```
to solve this:
1. origin_kind = SovereignAccount && white list `Treasurer` sovereign account for schedule.
2. origin_kind = Xcm && use something like `dispatch_as_account`

**Pull request closed without merging**